File: server/src/routes/matches.ts
import { Router, Request, Response } from 'express';
import db from '../db/database';

const router = Router();

// Shorten a placeholder string for the mobile short_name slot
function shortPlaceholder(text: string | null): string {
  if (!text) return 'TBD';
  const g = text.match(/^(\d)\. Gruppe ([A-L])$/);      if (g) return `${g[1]}.${g[2]}`;   // "1. Gruppe A" → "1.A"
  const s = text.match(/Sieger Spiel (\d+)/);            if (s) return `S.${s[1]}`;          // "Sieger Spiel 73" → "S.73"
  const v = text.match(/Verlierer Spiel (\d+)/);         if (v) return `V.${v[1]}`;          // "Verlierer Spiel 85" → "V.85"
  const b = text.match(/Bester 3\. \((\d)\)/);           if (b) return `B3.${b[1]}`;         // "Bester 3. (1)" → "B3.1"
  return text.substring(0, 5);
}

// GET /api/matches  ?group=A  &matchday=1  &status=scheduled  &phase=Achtelfinale
router.get('/', (req: Request, res: Response) => {
  const { group, matchday, status, phase } = req.query;
  const GROUP_NAMES = ['A','B','C','D','E','F','G','H','I','J','K','L'];

  let sql = `
    SELECT m.id, m.match_number, m.group_name, m.matchday,
           m.kickoff_utc, m.venue, m.home_goals, m.away_goals, m.status,
           m.home_placeholder, m.away_placeholder,
           ht.id AS home_team_id, ht.name AS home_team_name,
           ht.short_name AS home_short, ht.flag_emoji AS home_flag,
           at.id AS away_team_id, at.name AS away_team_name,
           at.short_name AS away_short, at.flag_emoji AS away_flag
    FROM   matches m
    LEFT JOIN teams ht ON ht.id = m.home_team_id
    LEFT JOIN teams at ON at.id = m.away_team_id
    WHERE  1=1
  `;
  const params: (string | number)[] = [];

  if (phase) {
    const p = String(phase);
    if (p === 'Vorrunde') {
      sql += ` AND m.group_name IN (${GROUP_NAMES.map(() => '?').join(',')})`;
      params.push(...GROUP_NAMES);
    } else {
      sql += ' AND m.group_name = ?';
      params.push(p);
    }
  }

  if (group)    { sql += ' AND m.group_name = ?'; params.push(String(group).toUpperCase()); }
  if (matchday) { sql += ' AND m.matchday = ?';   params.push(Number(matchday)); }
  if (status)   { sql += ' AND m.status = ?';     params.push(String(status)); }

  sql += ' ORDER BY m.kickoff_utc ASC, m.match_number ASC';

  const rows = db.prepare(sql).all(...params) as MatchRow[];
  res.json(rows.map(formatMatch));
});

// GET /api/matches/:id
router.get('/:id', (req: Request, res: Response) => {
  const row = db.prepare(`
    SELECT m.id, m.match_number, m.group_name, m.matchday,
           m.kickoff_utc, m.venue, m.home_goals, m.away_goals, m.status,
           m.home_placeholder, m.away_placeholder,
           ht.id AS home_team_id, ht.name AS home_team_name,
           ht.short_name AS home_short, ht.flag_emoji AS home_flag,
           at.id AS away_team_id, at.name AS away_team_name,
           at.short_name AS away_short, at.flag_emoji AS away_flag
    FROM   matches m
    LEFT JOIN teams ht ON ht.id = m.home_team_id
    LEFT JOIN teams at ON at.id = m.away_team_id
    WHERE  m.id = ?
  `).get(req.params.id);

  if (!row) { res.status(404).json({ error: 'Match not found' }); return; }
  res.json(formatMatch(row as MatchRow));
});

interface MatchRow {
  id: number; match_number: number; group_name: string; matchday: number;
  kickoff_utc: string; venue: string | null;
  home_goals: number | null; away_goals: number | null; status: string;
  home_placeholder: string | null; away_placeholder: string | null;
  home_team_id: number | null; home_team_name: string | null; home_short: string | null; home_flag: string | null;
  away_team_id: number | null; away_team_name: string | null; away_short: string | null; away_flag: string | null;
}

export function formatMatch(m: MatchRow) {
  // For KO matches where teams are not yet determined, use placeholder text as the team name
  const homeTeam = m.home_team_id != null
    ? { id: m.home_team_id, name: m.home_team_name!, short_name: m.home_short!, flag_emoji: m.home_flag }
    : { id: null,           name: m.home_placeholder ?? 'TBD', short_name: shortPlaceholder(m.home_placeholder), flag_emoji: null };

  const awayTeam = m.away_team_id != null
    ? { id: m.away_team_id, name: m.away_team_name!, short_name: m.away_short!, flag_emoji: m.away_flag }
    : { id: null,           name: m.away_placeholder ?? 'TBD', short_name: shortPlaceholder(m.away_placeholder), flag_emoji: null };

  return {
    id:               m.id,
    match_number:     m.match_number,
    group_name:       m.group_name,
    matchday:         m.matchday,
    kickoff_utc:      m.kickoff_utc,
    venue:            m.venue,
    status:           m.status,
    home_goals:       m.home_goals,
    away_goals:       m.away_goals,
    home_placeholder: m.home_placeholder,
    away_placeholder: m.away_placeholder,
    home_team:        homeTeam,
    away_team:        awayTeam,
  };
}

export default router;

File: server/routes/matches.py
"""Matches routes — equivalent of routes/matches.ts.

GET /api/matches           ?group=A &matchday=1 &status=scheduled &phase=Achtelfinale
GET /api/matches/:id
"""
import re
from fastapi import APIRouter, HTTPException, Query

from db.database import get_db

router = APIRouter()

_GROUP_STAGE = list("ABCDEFGHIJKL")


def _short_placeholder(text: str | None) -> str:
    """Generate a compact short_name from a KO placeholder string."""
    if not text:
        return "TBD"
    m = re.match(r"^(\d)\. Gruppe ([A-L])$", text)
    if m:
        return f"{m.group(1)}.{m.group(2)}"   # "1. Gruppe A" → "1.A"
    m = re.search(r"Sieger Spiel (\d+)", text)
    if m:
        return f"S.{m.group(1)}"              # "Sieger Spiel 73" → "S.73"
    m = re.search(r"Verlierer Spiel (\d+)", text)
    if m:
        return f"V.{m.group(1)}"              # "Verlierer Spiel 85" → "V.85"
    m = re.search(r"Bester 3\. \(([A-Z0-9/]+)\)", text)
    if m:
        return "B3."                           # "Bester 3. (A/B/C)" or "(1)" → "B3."
    return text[:5]


def _format_match(row) -> dict:
    """Convert a SQLite Row to a clean API dict."""
    home_team = (
        {
            "id": row["home_team_id"],
            "name": row["home_team_name"],
            "short_name": row["home_short"],
            "flag_emoji": row["home_flag"],
        }
        if row["home_team_id"] is not None
        else {
            "id": None,
            "name": row["home_placeholder"] or "TBD",
            "short_name": _short_placeholder(row["home_placeholder"]),
            "flag_emoji": None,
        }
    )
    away_team = (
        {
            "id": row["away_team_id"],
            "name": row["away_team_name"],
            "short_name": row["away_short"],
            "flag_emoji": row["away_flag"],
        }
        if row["away_team_id"] is not None
        else {
            "id": None,
            "name": row["away_placeholder"] or "TBD",
            "short_name": _short_placeholder(row["away_placeholder"]),
            "flag_emoji": None,
        }
    )
    return {
        "id": row["id"],
        "match_number": row["match_number"],
        "group_name": row["group_name"],
        "matchday": row["matchday"],
        "kickoff_utc": row["kickoff_utc"],
        "venue": row["venue"],
        "status": row["status"],
        "home_goals": row["home_goals"],
        "away_goals": row["away_goals"],
        "home_placeholder": row["home_placeholder"],
        "away_placeholder": row["away_placeholder"],
        "home_team": home_team,
        "away_team": away_team,
    }


_BASE_SQL = """
    SELECT m.id, m.match_number, m.group_name, m.matchday,
           m.kickoff_utc, m.venue, m.home_goals, m.away_goals, m.status,
           m.home_placeholder, m.away_placeholder,
           ht.id   AS home_team_id,   ht.name  AS home_team_name,
           ht.short_name AS home_short, ht.flag_emoji AS home_flag,
           at.id   AS away_team_id,   at.name  AS away_team_name,
           at.short_name AS away_short, at.flag_emoji AS away_flag
    FROM   matches m
    LEFT JOIN teams ht ON ht.id = m.home_team_id
    LEFT JOIN teams at ON at.id = m.away_team_id
    WHERE  1=1
"""


# ── GET /api/matches ──────────────────────────────────────────────────────────
@router.get("/")
def list_matches(
    group: str | None = Query(default=None),
    matchday: int | None = Query(default=None),
    status: str | None = Query(default=None),
    phase: str | None = Query(default=None),
):
    sql = _BASE_SQL
    params: list = []

    if phase:
        if phase == "Vorrunde":
            placeholders = ",".join("?" * len(_GROUP_STAGE))
            sql += f" AND m.group_name IN ({placeholders})"
            params.extend(_GROUP_STAGE)
        else:
            sql += " AND m.group_name = ?"
            params.append(phase)

    if group:
        sql += " AND m.group_name = ?"
        params.append(group.upper())
    if matchday is not None:
        sql += " AND m.matchday = ?"
        params.append(matchday)
    if status:
        sql += " AND m.status = ?"
        params.append(status)

    sql += " ORDER BY m.kickoff_utc ASC, m.match_number ASC"

    rows = get_db().execute(sql, params).fetchall()
    return [_format_match(r) for r in rows]


# ── GET /api/matches/:id ──────────────────────────────────────────────────────
@router.get("/{match_id}")
def get_match(match_id: int):
    row = get_db().execute(
        _BASE_SQL + " AND m.id = ?", (match_id,)
    ).fetchone()
    if not row:
        raise HTTPException(404, "Match not found")
    return _format_match(row)
